uvicorn/server.py

```diff
@@ -283,13 +298,38 @@ async def shutdown(self, sockets: Optional[List[socket.socket]] = None) -> None:
         if not self.force_exit:
             await self.lifespan.shutdown()
 
-    def install_signal_handlers(self) -> None:
+    @contextlib.contextmanager
+    def _capture_exit_signals(self) -> Iterator[None]:
         if threading.current_thread() is not threading.main_thread():
             # Signals can only be listened to from the main thread.
-            return
+            yield
+        else:
+            original_handlers = {sig: signal.getsignal(sig) for sig in HANDLED_SIGNALS}
+            self.install_signal_handlers()
+            yield
+            for sig, handler in original_handlers.items():
+                signal.signal(sig, handler)
+            # if we did gracefully shut down due to a signal, try to
+            # trigger the expected behaviour for this signal now
+            if self._delayed_signal is not None:
+                handler = original_handlers[self._delayed_signal]
+                # Python runs signal handlers in the main thread,
+                # so we just try to do the same first...
+                if callable(handler):
+                    handler(self._delayed_signal, None)
+                # ... if the signal is to be ignored anyway we do nothing...
+                elif handler == signal.SIG_IGN:
+                    pass
+                # ... if we are dealing with an internal/C signal handler,
+                #   repeat the signal to trigger the handler directly.
+                # Windows Caveat: We have no guarantee that we can correctly send
+                #   the signal (due to CTRL+C as SIGINT) to the process (due to groups)!
+                #   For now, do not do anything in this situation.
```

Comment:
Can we have some reference instead?

Reply:
I've added a reference to an Stack Overflow description of Windows signal handling and (re-)raising SIGINT/CTRL+C. Should I remove the other comments in the method?